apps/backend/src/datasources/stfc/StfcUserDataSource.ts

```diff
@@ -143,66 +146,92 @@ export class StfcUserDataSource implements UserDataSource {
       ? this.uowsSearchableBasicUserDetailsCache
       : this.uowsBasicUserDetailsCache;
 
-    const stfcUsers: StfcBasicPersonDetails[] = [];
+    const stfcUserRequests: Promise<StfcBasicPersonDetails | undefined>[] = [];
     const cacheMisses: string[] = [];
 
     for (const userNumber of userNumbers) {
       const cachedUser = cache.get(userNumber);
       if (cachedUser) {
-        stfcUsers.push(cachedUser);
+        stfcUserRequests.push(cachedUser);
       } else {
         cacheMisses.push(userNumber);
       }
     }
 
     if (cacheMisses.length > 0) {
-      const uowsRequest = searchableOnly
-        ? client.getSearchableBasicPeopleDetailsFromUserNumbers(
-            token,
-            cacheMisses
-          )
-        : client.getBasicPeopleDetailsFromUserNumbers(token, cacheMisses);
-      const usersFromUows: StfcBasicPersonDetails[] | null = (await uowsRequest)
-        ?.return;
+      const uowsRequest: Promise<StfcBasicPersonDetails[] | null> = (
+        searchableOnly
+          ? client.getSearchableBasicPeopleDetailsFromUserNumbers(
+              token,
+              cacheMisses
+            )
+          : client.getBasicPeopleDetailsFromUserNumbers(token, cacheMisses)
+      ).then((result) => result?.return);
+
+      for (const userNumber of cacheMisses) {
+        const userRequest = uowsRequest.then((users) =>
+          users?.find((user) => user.userNumber == userNumber)
+        );
+        cache.put(userNumber, userRequest);
+        stfcUserRequests.push(userRequest);
+      }
+
+      const usersFromUows = await uowsRequest;
 
       if (usersFromUows) {
         await this.ensureDummyUsersExist(
           usersFromUows.map((stfcUser) => parseInt(stfcUser.userNumber))
         );
-        usersFromUows.map((user) => cache.put(user.userNumber, user));
-        stfcUsers.push(...usersFromUows);
       }
     }
 
+    const stfcUsers: StfcBasicPersonDetails[] = await Promise.all(
+      stfcUserRequests
+    ).then((users) =>
+      users.filter((user): user is StfcBasicPersonDetails => user !== undefined)
+    );
+    // Uncache any failed lookups
+    userNumbers
+      .filter(
+        (un) => stfcUsers.find((user) => user.userNumber === un) === undefined
+      )
+      .forEach((un) => cache.remove(un));
+
     return stfcUsers;
   }
 
   private async getStfcBasicPersonByEmail(
     email: string,
     searchableOnly?: boolean
-  ): Promise<StfcBasicPersonDetails | null> {
+  ): Promise<StfcBasicPersonDetails | undefined> {
     const cache = searchableOnly
       ? this.uowsSearchableBasicUserDetailsCache
       : this.uowsBasicUserDetailsCache;
 
-    const cachedUser = cache.get(email);
+    const cachedUser = await cache.get(email);
     if (cachedUser) {
       return cachedUser;
```

Comment:
If I am understanding this correctly, then when since it cacheing the request and then returns this request. If that requests fails then are we not going to return a failed request until the cache times out? Sorry if I am miss understanding

Reply:
I was going for if the request fails, then resolving the cached Promise should return `undefined`, this return statement will be skipped, and we make a new request.

Reply:
Cool thanks just wanted to clarify that!